Convert gluon-announce Lua code to C modules

NeoRaider opened the pull request.

NeoRaider commented:
I'd like to merge the c-respondd ASAP. I've verified that its output is identical to the old code, and there don't seem to be any memory leaks.

jplitza commented on package/libgluon/src/libgluon.c:
```diff
@@ -0,0 +1,111 @@
+
+char * gluon_get_interface_address(const char *ifname) {
+        const char *format = "/sys/class/net/%s/address";
+        char path[strlen(format) + strlen(ifname)];
```
Shouldn't this be `strlen(format) + strlen(ifname) - 1`? The `%s` is replaced by `ifname` (-2) but `\0` is appended (+1).

NeoRaider replied:
You're right. I pondered about this a while when writing this, as it obviously doesn't matter if we reserve 1 byte more, but I should've at least added a comment. Thinking about this again, using the exact size is probably less confusing.

jplitza commented:
Some thoughts on package naming:
* libgluon is quite general and high-profile for some helper functions. Maybe libgluonutil or libgluon-util are better?
* gluon-announce now really is a misnomer, as it no longer contains any code related to announcing anything. Instead, it simply is a basic respondd provider like every other. The only thing special about it is that it provides the cache periods (which could maybe simply be included in gluon-announced instead?)
* gluon-announced could be renamed to gluon-respondd, to emphasize that it only starts respondd.

In fact, we can now probably merge gluon-announce and gluon-announced, as using the first without respondd is impossible, as is using respondd without the latter.

NeoRaider commented:
I agree that gluon-announce and gluon-announced can be merged to gluon-respondd.

I'm not sure about libgluon. libgluonutil is probably a better name, I'll rename it.

jplitza commented:
What does commit b13e233 do in this PR? But now that it is there: Didn't you object to stopping respondd during autoupdate?

NeoRaider commented:
@jplitza, well, it doesn't really belong in this PR, but it is based on the other patches; I think I'll remove it again before merging and merge it individually. (And Github shows the patch order incorrectly O_o).

No, I didn't object to stopping respondd at all.

jplitza commented:
Sorry, my memory betrayed me. I thought you wanted to have the node behave mostly normally during download, but that was with regard to everything *except* the mapping services.

Yes, Github shows the commits in a PR ordered randomly, it's really annoying.

jplitza commented on package/gluon-mesh-vpn-fastd/src/respondd.c:
```diff
@@ -0,0 +1,301 @@
+static struct json_object * get_peer_groups(struct json_object *groups, struct json_object *peers);
+
+static struct json_object * get_fastd_version(void) {
+        FILE *f = popen("exec fastd -v", "r");
```
Would it be too much work to do the fork ourselves and execute fastd directly, instead of spawning a shell in between?

NeoRaider replied:
I think it isn't worth the hassle...

jplitza commented on package/gluon-mesh-vpn-fastd/src/respondd.c:
```diff
@@ -0,0 +1,301 @@
+        pclose(f);
+
+        if (r >= 0) {
+                len = strlen(line);
```
I was confused, why you don't simply use the return value of `getline()`. Maybe add a comment explaining that?

jplitza commented on package/gluon-mesh-vpn-fastd/src/respondd.c:
```diff
@@ -0,0 +1,301 @@
+                line = NULL;
+        }
+
+        struct json_object *ret = gluonutil_wrap_string(line + 6);
```
I would like to check if `strncmp(line, "fastd ", 6) == 0` before stripping the first six bytes. If you want to do this unconditionally, at least add a comment that the first six bytes are "fastd ".

jplitza commented on package/gluon-mesh-vpn-fastd/src/respondd.c:
```diff
@@ -0,0 +1,301 @@
+static struct json_object * get_fastd(void) {
+        bool enabled = false;
+
+        struct uci_context *ctx = uci_alloc_context();
```
I'd propose packaging `uci_alloc_context()`, the setting of the strict flag and `uci_load()` into a libgluonutil function. It occurs four times in this PR, and especially the setting of the strict flag could easily be forgotten otherwise.

NeoRaider replied:
Could be done, but I think it's fine as is for now.

jplitza commented on package/gluon-node-info/src/respondd.c:
```diff
@@ -0,0 +1,144 @@
+
+        const char *role = get_first_option(ctx, p, "system", "role");
+        if (role && *role)
+                json_object_object_add(ret, "role", gluonutil_wrap_string(role));
```
Why does `get_system()` behave differently from `get_owner()` and `get_location()` in that it returns an empty JSON object instead of NULL when the corresponding setting isn't set?

NeoRaider replied:
This reflects the behaviour of the old Lua snippets. The `system` field should always exist, even when it is empty, while owner and location just don't exist when there's no data.

jplitza commented on package/gluon-respondd/files/etc/init.d/gluon-respondd:
```diff
@@ -0,0 +1,45 @@
+
+
+do_start() {
+	DEVS=$(cat $DEVLIST 2>/dev/null | while read dev iface; do echo -n " -i $dev"; done)
```
No need to use `cat` here, simply do `while read dev iface; do ...; done < $DEVLIST`

NeoRaider replied:
I think there's no way to hide the error message when $DEVLIST doesn't exist without the cat...

jplitza commented on package/gluon-respondd/files/etc/hotplug.d/iface/10-gluon-respondd:
```diff
@@ -0,0 +1,34 @@
+
+case "$ACTION" in
+	ifdown)
+		sed "/ $INTERFACE$/d" $DEVLIST > $DEVLIST.new
```
Why the temporary `.new` file instead of `sed -i`?

NeoRaider replied:
The specification of sed doesn't guarantee atomicity for `sed -i`. I think most (all?) current implementions use a temporary file for it, but making this explicit is cleaner.

jplitza commented on package/gluon-respondd/files/lib/gluon/respondd/neighbours.cache:
```diff
@@ -0,0 +1 @@
+10000
```
Didn't [you say](https://github.com/freifunk-gluon/packages/pull/125#discussion_r49883003) you want to change neighbours and statistics cache time to 900ms?

NeoRaider replied:
That's the plan, but I'd like to get some more real-life testing of the code before lowering the cache times, so this will be a post-2016.1 change.

jplitza commented on package/gluon-respondd/src/respondd.c:
```diff
@@ -0,0 +1,209 @@
+#include <sys/vfs.h>
+
+
+static struct json_object * gluonutil_version(void) {
```
Why is this function named `gluonutil_*` when it isn't in libgluonutil?

NeoRaider replied:
Oops. Too much search&replace.

jplitza commented on package/gluon-core/files/lib/gluon/upgrade/040-site-code:
```diff
@@ -0,0 +1,6 @@
+local sysconfig = require 'gluon.sysconfig'
+local site = require 'gluon.site_config'
+
+sysconfig.site_code = site.site_code
```
Where is this needed/where was this done before? It seems to be totally unrelated to everything else here.

NeoRaider replied:
This piece of code was added before we decided to convert site.conf to JSON (so creating a sysconfig file was the only way to get the site code from C). I forgot to remove it when rebasing...

jplitza commented on package/gluon-mesh-batman-adv-core/src/respondd.c:
```diff
@@ -0,0 +1,606 @@
+        snprintf(path, sizeof(path), format, ifname, name);
+
+        struct stat s;
+        return !stat(path, &s);
```
Why not use `!access(path, F_OK)` instead of `!stat(…)`? That wouldn't require an additional temporary struct.

jplitza commented on package/gluon-mesh-batman-adv-core/src/respondd.c:
```diff
@@ -0,0 +1,606 @@
+        struct json_object *other = json_object_new_array();
+
+        const char *format = "/sys/class/net/%s/lower_*";
+        char pattern[strlen(format) + strlen(ifname)];
```
[Again](https://github.com/freifunk-gluon/gluon/pull/612#discussion_r49765063), this should be `- 1`.

jplitza commented on package/gluon-mesh-batman-adv-core/files/lib/gluon/announce/nodeinfo.d/network/mesh_interfaces:
```diff
@@ -1,13 +0,0 @@
-  )
-end
-
-return interfaces
```
`nodeinfo.network.mesh_interfaces` isn't generated by the C version anymore. Obviously, the same information is contained in `nodeinfo.network.mesh.bat0.interfaces`, but in other form. Is this intended? Is it documented? Do we have something like API versions for what is returned by respondd?

NeoRaider replied:
`nodeinfo.network.mesh_interfaces` had been deprecated when `nodeinfo.network.mesh.bat0.interfaces` was added. This was before 2015.1, and the commit message mentioned that `nodeinfo.network.mesh_interfaces` would be removed after 2015.1.

At the moment, we don't have API versions or proper documentation of the API. We plan to switch to the well-documented [NetJSON](http://netjson.org/) format some time in the future (although we'd still have Gluon-specific extensions to NetJSON which would need to be documented...)

jplitza commented on package/gluon-mesh-batman-adv-core/src/respondd.c:
```diff
@@ -0,0 +1,606 @@
+        fclose(f);
+}
+
+static inline bool do_ioctl(int fd, struct ifreq *ifr, void *data) {
```
I think `do_ethtool_ioctl` would be a better name.

NeoRaider replied:
I'll add an `ethtool_` prefix to a few functions. I've copied the code from our lua-ethtool module we used so far, where ethtool prefixes didn't make much sense as everything was ethtool ;)

jplitza commented:
`package/gluon-mesh-batman-adv-core/src/respondd.c` has horribly mixed indentation, sometimes using spaces and sometimes tabs. All other `respondd.c` files use spaces exclusively, while respondd itself uses tabs. I could live with the discrepancy between the modules and the daemon, end even between the modules, but inside this same file, you really should settle with one of the options.

jplitza commented:
I have to admit, I currently don't have the leisure to dig into the documentation for iwinfo and ethtool, which means I didn't fully review the respondd.c from gluon-mesh-batman-adv-core yet. Also, I'm no good with Makefiles, so I didn't really review them either.

However, spontaneous tests with the compiled firmware looked good too. So modulo my comments, consider this ACK'ed.

NeoRaider commented:
The indentation issues were a consequence of me switching from Emacs to Atom recently, which I haven't fully mastered yet... I've changed everything to tabs, as it was intended.

Thanks again for your detailed review. I think I've addressed all your comments, so I'll merge it now after a last round of testing.
